File: letni25-26/aisd/pracownie/pracownia1/A.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <cstdio>

using namespace std;

int n_cities_iter(int n, int m, vector<vector<char>>& board) {
    
    auto is_in = [&](int i, int j) {
        return i >= 0 && j >= 0 && i < n && j < m;
    };

    auto remove_city = [&](int i, int j) {

        stack<pair<int,int>> st;
        st.push({i,j});

        while(!st.empty()) {
            pair<int,int> p = st.top();
            st.pop();
            int a = p.first;
            int b = p.second;

            char sq = board[a][b];

            if(sq == 'A') {
                continue;
            }
            else if(sq == 'B') {
                board[a][b] = 'A';
                if(is_in(a,b-1) && (board[a][b-1]=='D'||board[a][b-1]=='E'||board[a][b-1]=='F'))
                    st.push({a,b-1});
                if(is_in(a+1,b) && (board[a+1][b]=='C'||board[a+1][b]=='D'||board[a+1][b]=='F'))
                    st.push({a+1,b});
            }
            else if(sq == 'C') {
                board[a][b] = 'A';
                if(is_in(a,b-1) && (board[a][b-1]=='D'||board[a][b-1]=='E'||board[a][b-1]=='F'))
                    st.push({a,b-1});
                if(is_in(a-1,b) && (board[a-1][b]=='B'||board[a-1][b]=='E'||board[a-1][b]=='F'))
                    st.push({a-1,b});
            }
            else if(sq == 'D') {
                board[a][b] = 'A';
                if(is_in(a,b+1) && (board[a][b+1]=='B'||board[a][b+1]=='C'||board[a][b+1]=='F'))
                    st.push({a,b+1});
                if(is_in(a-1,b) && (board[a-1][b]=='B'||board[a-1][b]=='E'||board[a-1][b]=='F'))
                    st.push({a-1,b});
            }
            else if(sq == 'E') {
                board[a][b] = 'A';
                if(is_in(a,b+1) && (board[a][b+1]=='B'||board[a][b+1]=='C'||board[a][b+1]=='F'))
                    st.push({a,b+1});
                if(is_in(a+1,b) && (board[a+1][b]=='C'||board[a+1][b]=='D'||board[a+1][b]=='F'))
                    st.push({a+1,b});
            }
            else if(sq == 'F') {
                board[a][b] = 'A';
                if(is_in(a+1,b) && (board[a+1][b]=='C'||board[a+1][b]=='D'||board[a+1][b]=='F'))
                    st.push({a+1,b});
                if(is_in(a-1,b) && (board[a-1][b]=='B'||board[a-1][b]=='E'||board[a-1][b]=='F'))
                    st.push({a-1,b});
                if(is_in(a,b+1) && (board[a][b+1]=='B'||board[a][b+1]=='C'||board[a][b+1]=='F'))
                    st.push({a,b+1});
                if(is_in(a,b-1) && (board[a][b-1]=='D'||board[a][b-1]=='E'||board[a][b-1]=='F'))
                    st.push({a,b-1});
            }
        }
    };

    int cities = 0;

    for(int i=0;i<n;i++) {
        for(int j=0;j<m;j++) {
            if(board[i][j] != 'A') {
                cities++;
                remove_city(i,j);
            }
        }
    }

    return cities;
}


int main() {
    int n, m;

    scanf("%d %d\n", &n, &m);

    char buffer[m+2];

    vector<vector<char>> board(n, vector<char>(m));

    for(int i = 0; i < n; i++) {
        fgets(buffer, sizeof(buffer), stdin);
        for(int j = 0; j < m; j++) {
            board[i][j] = buffer[j];
        }
    }

    cout << n_cities_iter(n,m,board) << endl;
}

File: letni25-26/aisd/pracownie/pracownia1/A.py
def n_cities(n,m,board):
    #board is n x m
    def is_in(i,j):
        return i >= 0 and j >=  0 and i < n and j < m
    
    def remove_city(i,j):
        sq = board[i][j]
        if sq == 'A':
            return
        
        elif sq == 'B':
            board[i][j] = 'A'
            if is_in(i,j-1) and board[i][j-1] in ['D','E','F']:
                remove_city(i,j-1)
            if is_in(i+1,j) and board[i+1][j] in ['C','D','F']:
                remove_city(i+1,j)
        elif sq == 'C':
            board[i][j] = 'A'
            if is_in(i,j-1) and board[i][j-1] in ['D','E','F']:
                remove_city(i,j-1)
            if is_in(i-1,j) and board[i-1][j] in ['B','E','F']:
                remove_city(i-1,j)  
        elif sq == 'D':
            board[i][j] = 'A'
            if is_in(i,j+1) and board[i][j+1] in ['B','C','F']:
                remove_city(i,j+1)
            if is_in(i-1,j) and board[i-1][j] in ['B','E','F']:
                remove_city(i-1,j)
        elif sq == 'E':
            board[i][j] = 'A'
            if is_in(i,j+1) and board[i][j+1] in ['B','C','F']:
                remove_city(i,j+1)
            if is_in(i+1,j) and board[i+1][j] in ['C','D','F']:
                remove_city(i+1,j)
        elif sq == 'F':
            board[i][j] = 'A'
            if is_in(i+1,j) and board[i+1][j] in ['C','D','F']:
                remove_city(i+1,j)
            if is_in(i-1,j) and board[i-1][j] in ['B','E','F']:
                remove_city(i-1,j)
            if is_in(i,j+1) and board[i][j+1] in ['B','C','F']:
                remove_city(i,j+1)
            if is_in(i,j-1) and board[i][j-1] in ['D','E','F']:
                remove_city(i,j-1)

    cities = 0
    for i in range(n):
        for j in range(m):
            if board[i][j] != 'A':
                cities += 1
                print('removing: ',i,j)
                print(*board,sep='\n')
                remove_city(i,j)
            
    return cities

# with open('inputs.txt','r') as input:
#     input = input.read().splitlines()
#     n,m = input[0].split()
#     n,m = int(n),int(m)
#     board = [list(i) for i in input[1:]]
#     print(n,m,board)
#     print(n_cities(n,m,board))
#     print('after',board)


#wersja iteracyjna
class Stack:
    def __init__(self):
        self.stack = []
    def add(self,x):
        self.stack.append(x)
        return
    
    def pop(self):
        if self.stack == []:
            raise EnvironmentError
        x,rest = self.stack[-1],self.stack[:-1]
        self.stack = rest
        return x
    
    def is_empty(self):
        return self.stack == []
    
def n_cities_iter(n,m,board):
    #board is n x m
    def is_in(i,j):
        return i >= 0 and j >=  0 and i < n and j < m
    
    def remove_city(i,j):
        stack = Stack()
        
        stack.add((i,j))
        while not stack.is_empty():
            (a,b) = stack.pop()
            sq = board[a][b]
            if sq == 'A':
                continue
            
            elif sq == 'B':
                board[a][b] = 'A'
                if is_in(a,b-1) and board[a][b-1] in ['D','E','F']:
                    stack.add((a,b-1))
                if is_in(a+1,b) and board[a+1][b] in ['C','D','F']:
                    stack.add((a+1,b))
            elif sq == 'C':
                board[a][b] = 'A'
                if is_in(a,b-1) and board[a][b-1] in ['D','E','F']:
                    stack.add((a,b-1))
                if is_in(a-1,b) and board[a-1][b] in ['B','E','F']:
                    stack.add((a-1,b))  
            elif sq == 'D':
                board[a][b] = 'A'
                if is_in(a,b+1) and board[a][b+1] in ['B','C','F']:
                    stack.add((a,b+1))
                if is_in(a-1,b) and board[a-1][b] in ['B','E','F']:
                    stack.add((a-1,b))
            elif sq == 'E':
                board[a][b] = 'A'
                if is_in(a,b+1) and board[a][b+1] in ['B','C','F']:
                    stack.add((a,b+1))
                if is_in(a+1,b) and board[a+1][b] in ['C','D','F']:
                    stack.add((a+1,b))
            elif sq == 'F':
                board[a][b] = 'A'
                if is_in(a+1,b) and board[a+1][b] in ['C','D','F']:
                    stack.add((a+1,b))
                if is_in(a-1,b) and board[a-1][b] in ['B','E','F']:
                    stack.add((a-1,b))
                if is_in(a,b+1) and board[a][b+1] in ['B','C','F']:
                    stack.add((a,b+1))
                if is_in(a,b-1) and board[a][b-1] in ['D','E','F']:
                    stack.add((a,b-1))

    cities = 0
    for i in range(n):
        for j in range(m):
            if board[i][j] != 'A':
                cities += 1
                remove_city(i,j)
                
    return cities

with open('inputs.txt','r') as input:
    input = input.read().splitlines()
    n,m = input[0].split()
    n,m = int(n),int(m)
    board = [list(i) for i in input[1:]]
    print(n,m,board)
    print(n_cities_iter(n,m,board))
    print('after',board)